src/ngraph/op/generate_mask.cpp

```diff
@@ -0,0 +1,34 @@
+//*****************************************************************************
+// Copyright 2017-2018 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+
+#include "ngraph/op/generate_mask.hpp"
+
+using namespace std;
+using namespace ngraph;
+
+shared_ptr<Node> op::GenerateMask::copy_with_new_args(const NodeVector& new_args) const
+{
+    check_new_args_count(this, new_args);
+    return make_shared<GenerateMask>(
+        new_args.at(0), m_shape, m_element_type, m_seed, m_probability);
+}
+
+void ngraph::op::GenerateMask::validate_and_infer_types()
+{
+    NGRAPH_ASSERT(shape_size(get_training_node()->get_shape()) == 1)
```

Comment:
```suggestion
    NODE_VALIDATION_ASSERT(this, get_input_partial_type(0).compatible(PartialShape{}))
```

Any restrictions on the input element type?

Reply:
We should also check something like:

```
    NODE_VALIDATION_ASSERT(this, m_element_type.is_static())
        << "Output element type must not be dynamic.";
```

Reply:
I'm not sure why we want the "training flag" to be an input at all. Can't a bridge just omit `GenerateMask` if the mode is known statically, and replace it with something like `Select(is_training,GenerateMask,Constant(1))` if the mode is not known until runtime?

Reply:
This always generates the mask, even for inference, I belive? That gives us a slowdown. We're compiling separate inference and training graphs anyway, the selection happens at the graph level, not the op level.

Reply:
Then again, since I'm already compiling two graphs, I probably just won't insert the generate mask op into the graph if it's inference without randomization.

Reply:
Right, that’s what I was thinking (two graphs). If it really is a run-time decision I understand that it’s a problem though. I wish that we had a way to do something like if_else(cond,a,b) without forcing evaluation of both a and b, but we don’t.

It’s fine like it is in this respect, no need to block on this.